From f89a3d5ecc844731a1566a40224adc61b1f9a518 Mon Sep 17 00:00:00 2001
From: Roman Janota <Roman.Janota@cesnet.cz>
Date: Thu, 1 Oct 2026 17:32:54 +0200
Subject: [PATCH] sn BUGFIX on-change union filter descendants

The changes of an on-change subscription were iterated with "%s//.",
which for a union filter /a:x | /a:y means /a:x | (/a:y//.), so the
changes nested under every member but the last one were skipped. A
multi-sibling subtree filter converts into such a union.

The filter is now parenthesized before the descendant step.

Upstream: https://github.com/sysrepo/sysrepo/commit/6dc0ae0bd704132b3e2332733d4d6a5915449f34
Signed-off-by: Mattias Walström <lazzer@gmail.com>
---
 src/utils/sn_yang_push.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/utils/sn_yang_push.c b/src/utils/sn_yang_push.c
index a2e740b2..f61afcb8 100644
--- a/src/utils/sn_yang_push.c
+++ b/src/utils/sn_yang_push.c
@@ -567,9 +567,9 @@ srsn_yp_on_change_cb(sr_session_ctx_t *session, uint32_t UNUSED(sub_id), const c
 
     assert(sub->type == SRSN_YANG_PUSH_ON_CHANGE);
 
-    /* get the changes iterator */
+    /* get the changes iterator, the filter may be a union whose every member needs the descendants */
     if (xpath) {
-        r = asprintf(&xp, "%s//.", xpath);
+        r = asprintf(&xp, "(%s)//.", xpath);
     } else {
         r = asprintf(&xp, "/%s:*//.", module_name);
     }
